include/ringbuffer_utils.hpp

```diff
@@ -152,6 +153,151 @@ class PerfRingBufferReader {
   uint64_t _head;
 };
 
+struct MPSCRingBufferHeader {
+  uint64_t size;
+  static constexpr uint64_t k_discard_bit = 1UL << 62;
+  static constexpr uint64_t k_busy_bit = 1UL << 63;
+
+  static bool is_busy(uint64_t size) { return size & k_busy_bit; }
+  static bool is_discarded(uint64_t size) { return size & k_discard_bit; }
+};
+
+class MPSCRingBufferWriter {
+public:
+  explicit MPSCRingBufferWriter(RingBuffer &rb) : _rb(rb) {
+    assert(rb.type == RingBufferType::kMPSCRingBuffer);
+    _tail = __atomic_load_n(_rb.reader_pos, __ATOMIC_ACQUIRE);
+  }
+
+  MPSCRingBufferWriter(const MPSCRingBufferWriter &) = delete;
+  MPSCRingBufferWriter &operator=(const MPSCRingBufferWriter &) = delete;
+
+  void update_tail() {
+    _tail = __atomic_load_n(_rb.reader_pos, __ATOMIC_ACQUIRE);
+  }
+
+  Buffer reserve(size_t n) {
```

Comment:
I really like `reserve()` semantics here.

One of the challenges though is that a consumer doesn't know if a given element is ready, only if the whole ringbuffer is ready.  I think the current strategy for that problem is good, since the alternative is to have a bitmask for potential reader positions indicating whether the given read position is ready (since we have variable-size elements).  I don't think that's a favorable alternative unless we start hitting issues with sparse updates (lagging thread in the middle of the ringbuffer) or high write contention with fast drains (in which case the buffer is never full but always busy), but I very much doubt that will happen given how big our samples are.

Reply:
I would be interested to discuss your idea !
Initially I started an implementation without a spinlock, the idea was to use a ready bit, at the start the ready bit would be unset, writer would reserve memory by advancing head through a CAS operation (as it is done here), but reader would not read the sample since its ready is not set yet. Upon commit writer would do an exchange on the ready bit to set it, and reader would do an exchange on it with 0 to check if sample is ready and at the same time reset it to not ready when ready. Sadly this does not work with variable-size samples :(
The other idea was to have a "pre-head" that ahead of head, reserving would advance the pre-head through CAS. Here there is not need for ready/busy state, but commit is more complex because only the writer that has reserved the sample just after head is allowed to advance head, which leads to a wait loop.